### PR TITLE
reparent: flush binary logs on candidate primary

### PR DESCRIPTION
**Pull request by deepthi**
## Description

The problem described in #4161 has resurfaced. This is not actually a regression, it seems like the fix in #4179 doesn't necessarily help in a situation where the binlog file is large and the cluster is taking a lot of write traffic.

This PR tries to address that by flushing the binary logs **before** demoting the original primary. The problem with doing it **after** is that if no writes have happened after the binlog is flushed, then you still need to scan the previous file to find the GTID you are looking for.

The idea here is that in a high-traffic system, you are almost guaranteed to have some writes between flushing the binlogs and demoting the primary. In a low-traffic system it doesn't really matter.

## Related Issue(s)

#4161 #4163 #4179

## Checklist
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/mysqlctl/reparent.go [hunk 1]
```diff
@@ -118,3 +118,24 @@ func (mysqld *Mysqld) Promote(hookExtraEnv map[string]string) (mysql.Position, e
...
+	defer conn.Recycle()
+
+	// Since we handle replication, just stop it.
+	cmds := []string{
+		// When using semi-sync and GTID, a replica first connects to the new master with a given GTID set,
+		// it can take a long time to scan the current binlog file to find the corresponding position.
+		// This can cause commits that occur soon after the master is promoted to take a long time waiting
+		// for a semi-sync ACK, since replication is not fully set up.
+		// More details in: https://github.com/vitessio/vitess/issues/4161
+		"FLUSH BINARY LOGS",
```
Comment (deepthi) [inline go/vt/mysqlctl/reparent.go:137]: @rafael do you think we should delete this from `Promote`? Lines 108-113.

**Comment by rafael** [comment]
I'm still puzzled by this issue. 

>The problem with doing it after is that if no writes have happened after the binlog is flushed, then you still need to scan the previous file to find the GTID you are looking for.
If the goal is to solve this for high write throughput systems, it seems that this shouldn't be a concern. 

I spoke with @aquarapid about this issue last week and I tried to reproduce this problem in a system with large binlogs and also high throughput writes. I couldn't reproduce and the existent solution was sufficient. I also tested reverting the original PR and ran into issues where PlannedReparents took a very long time. 

One more thought. I don't love the idea of having to add one more rpc call in the PlannedReparent codepath. This is something that we also tried to avoid the first time we did this. I was thinking that another way of handling this issue is by recommending users to tweak `max_binlog_size` so the binlog files never grow too large. I'm curious if this will be sufficient.

**Comment by aquarapid** [comment]
I have reproduced the issue, to some extent.  I could get up to PRS up 20-30 seconds failover time (instead of the usual < 2-3 secs), not quite enough to time the PRS out;  but that may have been due to the shape of my workload, storage IO, etc.

I am not a huge fan of recommending a max_binlog_size change, since keeping it large is recommended by many MySQL experts, and I think we want a "good" experience regardless?

File: go/vt/mysqlctl/reparent.go (hunk 1 above)
Comment (rafael) [reply go/vt/mysqlctl/reparent.go:137]: Yup. If we decide with this new approach, we should remove this.

**Comment by rafael** [comment]
> I have reproduced the issue, to some extent. I could get up to PRS up 20-30 seconds failover time (instead of the usual < 2-3 secs), not quite enough to time the PRS out; but that may have been due to the shape of my workload, storage IO, etc.
>
Were you able to reproduce in MySQL 5.7?

> I am not a huge fan of recommending a max_binlog_size change, since keeping it large is recommended by many MySQL experts,

Do you know what's the rationale behind that recommendation? For different reasons, we flush binary logs regularly and our binlog files are relatively small (below 100mb). We haven't seen any issues with this.

>and I think we want a "good" experience regardless?

Yeah I share this as well.

@deepthi  - If this is a burning issue and the change here fixes the problem you are seeing, I don't have objections going with this approach. Expanding on why I'm hesitant: 

a) I couldn't reproduce this problem with the same conditions that the PR mentions will fix. I worry we a re missing something. This is the setup that I had:

1) Test table: 
```
CREATE TABLE `test_table` (
  `i` int(11) NOT NULL,
  `c` char(10) DEFAULT NULL,
  `counter` int(11) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4
```

2) Generating 6K QPS inserts like the following:
```
INSERT INTO test_table(i) VALUES (1)
```
3) Let the binlog files grow to 4GB of data.

4) Perform a PlannedReparent and assert that it finishes within few seconds. 

@aquarapid  - Is this test similar to the one you are doing? Could you try a similar test in your end and check if you get the same results as I did? 

b)  It seems we will be using a somewhat big hammer (the new sync rpc call in PRS), for something that is ultimately a MySQL bug.

**Comment by aquarapid** [comment]
@rafael my test is similar;  but more specifically;  I let the current binlog file grow until it's near it's per file maximum (which is 1 GB; 
 the mysql default, and also maximum I believe).

RE the MySQL binlog size recommendation;  this is the MySQL default, and many DBAs are resistant to change it because they don't want to deal with the increase in number of binlogs that goes with this.  There's also the perception that switch binlogs has a performance impact when it occurs, something that is probably not true anymore, but has been in the past.

**Comment by deepthi** [comment]
We now think the root cause was errant gtids.

**Pull request closed without merging**